### DISCUSSION TITLE
ELRS Relay/repeater questions (protocol, inverters, CRSF addresses, frequencies etc)

### DISCUSSION THREAD

**Discussion by kolins-cz** [ExpressLRS/ExpressLRS#2201] [Q&A]
![relay2](https://user-images.githubusercontent.com/8409715/236402722-bd5ef522-5bf1-4fc4-926b-12082eae473d.png)
![sport to uart](https://user-images.githubusercontent.com/8409715/236402732-a46e2da9-1122-47a2-8390-c01bf173d7ad.png)

Hello,
As many asked before me I want to make a relay with two ELRS systems. But there are still no real answers. (like this dead thread https://github.com/ExpressLRS/ExpressLRS/discussions/1450)
If you take a look at the diagram, everything looks fine in theory, but will it work? 

- Would it be OK to use 2.4GHz all the way?
- I have discovered that I would need ESP32 based module at **3** as it can understand S.PORT directly without inverters and diode via single wire. Frsky R9M module has inverters and diode integrated so external converter (as seen on my schematic) needs to be used. Correct?
- Obviously I am interested in telemetry from remote link. Would getting telemetry trough short link need special support in firmware of **1** and **3**? If this would be hard to solve problem I can live without telemetry on remote link. 
- TBS CRSF documentation states addressing in CRSF protocol. Not sure but I would bet that ELRS receivers spits out packets with address 0xEA (Radio Transmitter) or 0xC8 (Flight controller). Unit **3** might ignore packets send by **2** as they are not addressed for it?
- Do I really need some MCU in between **2** and **3** as a man-in-the-middle to alter CRSF packets? There is one project on github for this, but poorly documented.
- What about failsafe? If short link is lost, remote link needs to generate failsafe immediately otherwise it would lead to flyaway.
- Can **3** and **4** be a TBS Crossfire system? 
- Is there some tool to eavesdrop CRSF and to see what is going on? I have tried to look on raw CRSF protocol in HEX but it is not trivial as it has variable length and many options.
- Is there more recent version of CRSF protocol documentation? Old one: [CRSF Rev07.pdf](https://github.com/ExpressLRS/ExpressLRS/files/11404179/CRSF.Rev07.pdf) Frankly this protocol description seems not so great. I have studied and implemented other protocols with better documentation. Maybe newer version will be better.

My goal is to build a relay / repeater without altering ELRS firmware (much). I have done RC repeater before, in days of 35MHz FM and OpenLRS. But it was trivial with PPM all the way. 
There are projects to convert SBUS to CRSF like this one [https://github.com/qubic/sbus2expresslrs](url) but I would prefer to use ELRS on both links.

**Comment by kielerrr** [comment]
I'm also using the same Rev07 doc to reference CRSF. It's 6 years old and was geared for TBS. From what I can tell, the doc is valid but is missing a lot of message of destinations and types. 
The codebase has a few header files for CRSF that give a better picture. [crsf_protocol.h](https://github.com/ExpressLRS/ExpressLRS/blob/257af811084497b24da03f32db0f0293b185bbb7/src/lib/CrsfProtocol/crsf_protocol.h ) is a good one.

To read packets on the RX, attach a USB/TTL device to the RX. I use `cutecom` to open the USB/TTL serial device hooked to the RX and if you have baud set correctly you'll see the raw CRSF packets in hex. Also check if your USB/TTL device has rx/tx pins inverted or not.
In that wave of bits coming across the wire, you'll see it's just repeating patterns.. If you look closer you'll see the repeating patterns start with an `ee` or a `c8` which is a sync bit indicating if the packet is going to the RX or the TX etc.. But if you see those patterns in the serial stream you should be able to read the bits.

Edit: 
Also, instead of having to use JR pins, if the device has USB, you can go to http://elrs_tx.local/hardware.html and the first two fields let you set the rx/tx pins. On the Radiomaster Ranger, the default rx/tx pins are set to 13 which are the JR pins. Changing that to the USB pins lets you hookup to USB. On the Ranger, you change rx/tx from 13 to 3 and 1 respectively.

**Reply by yuriy-boyun** [reply]
@kolins-cz were you able to make relay work with UART-SBUS converter?

**Comment by agoliveira** [comment]
With ELRS 3.3, the receiver can use SBUS but lacks telemetry. I did work in a project that uses SBUS+SPORT but would be nice if ELRS supported FPORT

**Reply by frobert0** [reply]
Hello, what is your project, can you send me info, please thank you.
Just starting with it, would be very interested

**Comment by kolins-cz** [comment]
I ha stumbled upon this commit, does this mean that relay is supposed to work by now?
https://github.com/ExpressLRS/ExpressLRS/commit/a921e606201bec8ab65c58140d6f977441e4fd62

**Comment by omgzhyla** [comment]
@kolins-cz have you succeeded in building that relay?

**Comment by Levimaxx** [comment]
hi, did it work? If it does, I want to know how you make it. I want to study but i don't have money to buy many modules. I what to know exactly. Thank you so much for your previous conversation.
@kolins-cz

**Comment by IgnacioHR** [comment]
Now that we have Gemini, wouldn’t be easier to receive in one frequency and transmit in the other frequency? so the receiver doesn’t perceive any change in anything? Maybe I’m just ignoring important things. The repeater should bind as another receiver and it could capture the two frequencies the gemini is using and decide to ear in one frequency and retransmit in the other one after measuring RSSI in both frequencies … thoughts?

**Comment by evgeny17387new** [comment]
hi @kolins-cz , have you got it working ? i am also trying to build one by myself

**Reply by frobert0** [reply]
Hello, how far have you come, can you send me info, please thank you.
Just starting with it, would be very interested

**Comment by flybabo** [comment]
It would be great if we can configure an ELRS Rx as a single wire inverted half-duplex CRSF mode (repeater mode?) just like an ELRS Tx module communicating with a S.PORT pin on a transmitter module bay.

**Comment by wimalopaan** [comment]
https://github.com/ExpressLRS/ExpressLRS/issues/3289

**Comment by wimalopaan** [comment]
I made a simple CRSF-package-rewriter for that purpose.
